tools/integration-tests/tester/tests/synchronization_test.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/iotaledger/goshimmer/tools/integration-tests/tester/framework"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestNodeSynchronization checks whether messages are synchronized by a peer that joined the network later
+// and initially missed some messages.
+func TestNodeSynchronization(t *testing.T) {
+	n, err := f.CreateNetwork("TestNodeSynchronization", 4, 2)
+	require.NoError(t, err)
+	defer n.Shutdown()
+
+	numMessages := 100
+	idsMap := make(map[string]MessageSent, numMessages)
+	ids := make([]string, numMessages)
+
+	// create messages on random peers
+	for i := 0; i < numMessages; i++ {
+		data := []byte(fmt.Sprintf("Test%d", i))
+
+		peer := n.RandomPeer()
+		id, err := peer.Data(data)
+		require.NoError(t, err)
+
+		idsMap[id] = MessageSent{
+			number: i,
+			id:     id,
+			// save payload to be able to compare API response
+			data:            payload.NewData(data).Bytes(),
+			issuerPublicKey: peer.Identity.PublicKey().String(),
+		}
+		ids[i] = id
+	}
+
+	// wait for messages to be gossiped
+	time.Sleep(5 * time.Second)
+
+	// make sure every peer got every message
+	checkForMessageIds(t, n.Peers(), numMessages, ids, idsMap)
+
+	// spawn peer without knowledge of previous messages
+	newPeer, err := n.CreatePeer()
+	require.NoError(t, err)
+	err = n.WaitForAutopeering(3)
+	require.NoError(t, err)
+
+	ids2 := make([]string, numMessages)
+
+	// create messages on random peers
+	for i := 0; i < numMessages; i++ {
+		data := []byte(fmt.Sprintf("Test%d", i))
+
+		peer := n.RandomPeer()
+		id, err := peer.Data(data)
+		require.NoError(t, err)
+
+		ids2[i] = id
+		idsMap[id] = MessageSent{
+			number: i,
+			id:     id,
+			// save payload to be able to compare API response
+			data:            payload.NewData(data).Bytes(),
+			issuerPublicKey: peer.Identity.PublicKey().String(),
+		}
+	}
+
+	// wait for messages to be gossiped
+	time.Sleep(5 * time.Second)
+
+	// check whether peer has synchronized ids (previous messages)
+	checkForMessageIds(t, []*framework.Peer{newPeer}, numMessages, ids, idsMap)
+
+	// make sure every peer got every message
+	checkForMessageIds(t, n.Peers(), numMessages, ids2, idsMap)
+}
+
+type MessageSent struct {
+	number          int
+	id              string
+	data            []byte
+	issuerPublicKey string
+}
+
+func checkForMessageIds(t *testing.T, peers []*framework.Peer, numMessages int, ids []string, idsMap map[string]MessageSent) {
+	for _, peer := range peers {
+		resp, err := peer.FindMessageByID(ids)
+		require.NoError(t, err)
+
+		// check for the count of messages
+		assert.Equal(t, numMessages, len(resp.Messages))
+
+		// check that all messages are present in response
+	outer:
+		for _, id := range ids {
+			for _, msg := range resp.Messages {
+				// if message found skip to next
+				if msg.ID == id {
+					continue outer
+				}
+			}
+
+			t.Errorf("MessageId=%s, issuer=%s not found in peer %s.", id, idsMap[id].issuerPublicKey, peer.String())
+		}
```

Comment:
```suggestion
		respIDs := make(map[string]struct{}, len(resp.Messages))
		for _, msg := range resp.Messages {
			respIDs[msg.ID] = struct{}{}
		}
		for _, id := range ids {
			assert.Containsf(t, respIDs, id, "MessageId=%s, issuer=%s not found in peer %s.", id, idsMap[id].issuerPublicKey, peer.String())
		}
```
I guess this is also what @luca-moser was referring to earlier.
IMHO it looks a bit nicer (by avoiding the label) and hash linear runtime instead of quadratic (not that this matters much in this test, though)

Reply:
I changed it to a similar approach but at the same time making sure that `resp.Messages` also contains no additional `ids`.

Reply:
also in the above you already do that with `assert.Equal(t, numMessages, len(resp.Messages))`

Reply:
There is another option if you want to assure that in one go:
```go
respIDs := make([]string, len(resp.Messages))
for i, msg := range resp.Messages {
	respIDs[i] = msg.ID
}
assert.ElementsMatch(t, ids, respIDs)
```
That is definitely the shortest version and we let `assert` handle all the interesting stuff.

Reply:
Definitely the most elegant solution. However, it doesn't yield as much information if a message is missing. But I think this can be inferred from the logs anyway.